Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp

```diff
@@ -104,21 +105,134 @@ std::optional<RTCRtpCapabilities> PeerConnectionBackend::senderCapabilities(Scri
+};
+
+JSONFileHandler& jsonFileHandler(const String& path)
+{
+    static NeverDestroyed<JSONFileHandler> sharedInstance(path);
+    return sharedInstance;
+}
+#endif
+
 PeerConnectionBackend::PeerConnectionBackend(RTCPeerConnection& peerConnection)
     : m_peerConnection(peerConnection)
 #if !RELEASE_LOG_DISABLED
     , m_logger(peerConnection.logger())
     , m_logIdentifier(peerConnection.logIdentifier())
+    , m_logIdentifierString(makeString(hex(m_logIdentifier)))
 #endif
 {
 #if USE(LIBWEBRTC)
     RefPtr document = peerConnection.document();
     if (auto* page = document ? document->page() : nullptr)
         m_shouldFilterICECandidates = page->webRTCProvider().isSupportingMDNS();
 #endif
+
+#if !RELEASE_LOG_DISABLED && (PLATFORM(WPE) || PLATFORM(GTK))
+    m_jsonFilePath = String::fromUTF8(getenv("WEBKIT_WEBRTC_JSON_EVENTS_FILE"));
+    if (!m_jsonFilePath.isEmpty())
+        jsonFileHandler(m_jsonFilePath).addClient(m_logIdentifier);
+#endif
+
+#if !RELEASE_LOG_DISABLED
+    m_logger->addObserver(*this);
+    ALWAYS_LOG(LOGIDENTIFIER, "PeerConnection created"_s);
+#endif
+}
+
+PeerConnectionBackend::~PeerConnectionBackend()
+{
+#if !RELEASE_LOG_DISABLED
+    ALWAYS_LOG(LOGIDENTIFIER, "Disposing PeerConnection"_s);
+    m_logger->removeObserver(*this);
+#endif
+
+#if !RELEASE_LOG_DISABLED && (PLATFORM(WPE) || PLATFORM(GTK))
+    if (!m_jsonFilePath.isEmpty())
+        jsonFileHandler(m_jsonFilePath).removeClient(m_logIdentifier);
+#endif
 }
 
-PeerConnectionBackend::~PeerConnectionBackend() = default;
+#if !RELEASE_LOG_DISABLED
+bool PeerConnectionBackend::handleLogMessage(const WTFLogChannel& channel, WTFLogLevel, Vector<JSONLogValue>&& values)
+{
+#if PLATFORM(WPE) || PLATFORM(GTK)
+    auto name = StringView::fromLatin1(channel.name);
+    if (name != "WebRTC"_s)
+        return false;
+
+    // Ignore logs containing only the call site information or JSON logs.
+    if (values.size() < 2 || values[1].type == JSONLogValue::Type::JSON)
+        return false;
+
+    // Parse "foo::bar(hexidentifier) "
+    auto& callSite = values[0].value;
+    auto leftParenthesisIndex = callSite.reverseFind('(');
+    if (leftParenthesisIndex == notFound)
+        return false;
+
+    auto rightParenthesisIndex = callSite.reverseFind(')');
+    if (rightParenthesisIndex == notFound)
+        return false;
+
+    auto identifier = callSite.substring(leftParenthesisIndex + 1, rightParenthesisIndex - leftParenthesisIndex - 1);
+    if (identifier != m_logIdentifierString)
+        return false;
+
+    if (!isJSONLogStreamingEnabled())
+        return false;
+
+    // Assume the third message is a multi-lines payload (such as SDP).
+    std::optional<std::span<const uint8_t>> payloadValue;
+    if (values.size() == 3)
+        payloadValue = values[2].value.span8();
+
+    auto event = generateJSONLogEvent(MessageLogEvent { values[1].value, WTFMove(payloadValue) }, false);
+    emitJSONLogEvent(WTFMove(event));
```

Comment:
Why not passing a MessageLogEvent struct instead of a string?
And maybe an enumeration value instead of a bool?

Reply:
`provideStatsLogs()` takes a String.